File: apps/ios/Orange Cloud/Orange Cloud/ViewModels/ZonePerformanceViewModel.swift
//
//  ZonePerformanceViewModel.swift
//  Orange Cloud
//
//  Zone 详情页「性能与缓存」面板：网络优化开关（Brotli / HTTP2 / HTTP3 / 0-RTT /
//  Early Hints / WebSockets / IPv6）+ 缓存控制（缓存级别 / Always Online / 查询字符串排序）。
//  全部走通用 zone 设置端点，读 zone-settings.read，写 zone-settings.write。
//

import Foundation
import Observation

/// 缓存级别（zone setting `cache_level` 的取值）
nonisolated enum CacheLevel: String, CaseIterable, Identifiable, Sendable {
    case basic, simplified, aggressive

    var id: String { rawValue }

    var title: String {
        switch self {
        case .basic:      String(localized: "无查询字符串")
        case .simplified: String(localized: "忽略查询字符串")
        case .aggressive: String(localized: "标准")
        }
    }
}

@Observable
@MainActor
final class ZonePerformanceViewModel {

    /// 网络优化里的 on/off 设置，按 Cloudflare 仪表盘顺序
    static let networkToggles = ["brotli", "http2", "http3", "0rtt", "early_hints", "websockets", "ipv6"]

    private(set) var values: [String: String] = [:]
    private(set) var loaded = false
    private(set) var isLoading = false
    /// 正在写入的 setting ID
    var updating: Set<String> = []
    var error: String?

    private let service: ZoneSettingsService
    private let zoneId: String

    init(service: ZoneSettingsService, zoneId: String) {
        self.service = service
        self.zoneId = zoneId
    }

    func isOn(_ id: String) -> Bool { values[id] == "on" }
    var cacheLevel: CacheLevel { CacheLevel(rawValue: values["cache_level"] ?? "") ?? .aggressive }

    func load() async {
        guard !loaded, !isLoading else { return }
        isLoading = true
        defer { isLoading = false }

        async let brotli      = service.getSetting(zoneId: zoneId, setting: "brotli")
        async let http2       = service.getSetting(zoneId: zoneId, setting: "http2")
        async let http3       = service.getSetting(zoneId: zoneId, setting: "http3")
        async let zeroRTT     = service.getSetting(zoneId: zoneId, setting: "0rtt")
        async let earlyHints  = service.getSetting(zoneId: zoneId, setting: "early_hints")
        async let websockets  = service.getSetting(zoneId: zoneId, setting: "websockets")
        async let ipv6        = service.getSetting(zoneId: zoneId, setting: "ipv6")
        async let cacheLevel  = service.getSetting(zoneId: zoneId, setting: "cache_level")
        async let alwaysOnline = service.getSetting(zoneId: zoneId, setting: "always_online")
        async let sortQS      = service.getSetting(zoneId: zoneId, setting: "sort_query_string_for_cache")

        var acc: [String: String] = [:]
        if let v = try? await brotli       { acc["brotli"] = v }
        if let v = try? await http2        { acc["http2"] = v }
        if let v = try? await http3        { acc["http3"] = v }
        if let v = try? await zeroRTT      { acc["0rtt"] = v }
        if let v = try? await earlyHints   { acc["early_hints"] = v }
        if let v = try? await websockets   { acc["websockets"] = v }
        if let v = try? await ipv6         { acc["ipv6"] = v }
        if let v = try? await cacheLevel   { acc["cache_level"] = v }
        if let v = try? await alwaysOnline { acc["always_online"] = v }
        if let v = try? await sortQS       { acc["sort_query_string_for_cache"] = v }

        // 一项都读不到（通常是无 zone-settings.read）→ 保持未加载态，UI 显示锁定
        guard !acc.isEmpty else { return }
        values = acc
        loaded = true
    }

    func setToggle(_ id: String, _ on: Bool) async {
        await update(id, value: on ? "on" : "off")
    }

    func setCacheLevel(_ level: CacheLevel) async {
        await update("cache_level", value: level.rawValue)
    }

    private func update(_ id: String, value: String) async {
        guard !updating.contains(id) else { return }
        updating.insert(id)
        error = nil
        do {
            let applied = try await service.setSetting(zoneId: zoneId, setting: id, value: value)
            values[id] = applied
        } catch {
            self.error = error.localizedDescription
        }
        updating.remove(id)
    }
}

File: apps/android/app/src/main/kotlin/jiamin/chen/orangecloud/ui/zonesettings/ZonePerformanceViewModel.kt
package jiamin.chen.orangecloud.ui.zonesettings

import androidx.annotation.StringRes
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import jiamin.chen.orangecloud.R
import jiamin.chen.orangecloud.core.auth.AuthRepository
import jiamin.chen.orangecloud.core.auth.Scopes
import jiamin.chen.orangecloud.data.repository.ZoneSettingsRepository
import kotlinx.coroutines.async
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.receiveAsFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

/** 缓存级别（zone setting `cache_level` 取值）。对应 iOS CacheLevel。 */
enum class CacheLevel(val raw: String, @StringRes val titleRes: Int) {
    BASIC("basic", R.string.perf_cache_basic),
    SIMPLIFIED("simplified", R.string.perf_cache_simplified),
    AGGRESSIVE("aggressive", R.string.perf_cache_aggressive);

    companion object {
        fun fromRaw(raw: String?): CacheLevel = entries.firstOrNull { it.raw == raw } ?: AGGRESSIVE
    }
}

/** 一个网络优化开关：setting id + 显示名。按 Cloudflare 仪表盘顺序。 */
data class PerfToggle(val id: String, @StringRes val labelRes: Int)

data class ZonePerformanceUiState(
    val zoneName: String = "",
    /** setting id → 当前值（"on"/"off" 或 cache_level 取值）。读不到的项不在 map 中。 */
    val values: Map<String, String> = emptyMap(),
    val loaded: Boolean = false,
    val isLoading: Boolean = false,
    val missingScope: Boolean = false,
    val canWrite: Boolean = false,
    val updating: Set<String> = emptySet(),
) {
    fun isOn(id: String): Boolean = values[id] == "on"
    val cacheLevel: CacheLevel get() = CacheLevel.fromRaw(values["cache_level"])
}

/**
 * 「性能与缓存」面板：网络优化开关 + 缓存控制。全部走 zone-settings（对应 iOS ZonePerformanceViewModel）。
 */
@HiltViewModel
class ZonePerformanceViewModel @Inject constructor(
    savedStateHandle: SavedStateHandle,
    private val repository: ZoneSettingsRepository,
    authRepository: AuthRepository,
) : ViewModel() {

    private val zoneId: String = checkNotNull(savedStateHandle["zoneId"])
    private val hasRead = authRepository.hasScope(Scopes.ZONE_SETTINGS_READ)
    private val canWrite = authRepository.hasScope(Scopes.ZONE_SETTINGS_WRITE)

    private val _uiState = MutableStateFlow(
        ZonePerformanceUiState(
            zoneName = savedStateHandle.get<String>("zoneName").orEmpty(),
            isLoading = hasRead,
            missingScope = !hasRead,
            canWrite = canWrite,
        ),
    )
    val uiState: StateFlow<ZonePerformanceUiState> = _uiState.asStateFlow()

    private val eventChannel = Channel<String?>(Channel.BUFFERED)
    val errors: Flow<String?> = eventChannel.receiveAsFlow()

    init {
        if (hasRead) load()
    }

    fun load() {
        if (!hasRead) return
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true) }
            try {
                // 每项独立读取，单项失败（如某些套餐无该设置）不影响其它。
                val ids = NETWORK_TOGGLES.map { it.id } + listOf("cache_level", "always_online", "sort_query_string_for_cache")
                val results = ids.map { id ->
                    id to async { runCatching { repository.getSetting(zoneId, id) }.getOrNull() }
                }
                val acc = buildMap {
                    results.forEach { (id, deferred) -> deferred.await()?.let { put(id, it) } }
                }
                if (acc.isNotEmpty()) {
                    _uiState.update { it.copy(values = acc, loaded = true) }
                }
            } finally {
                _uiState.update { it.copy(isLoading = false) }
            }
        }
    }

    fun setToggle(id: String, on: Boolean) = update(id, if (on) "on" else "off")

    fun setCacheLevel(level: CacheLevel) = update("cache_level", level.raw)

    private fun update(id: String, value: String) {
        if (!canWrite || id in _uiState.value.updating) return
        _uiState.update { it.copy(updating = it.updating + id) }
        viewModelScope.launch {
            try {
                val applied = repository.setSetting(zoneId, id, value)
                _uiState.update { it.copy(values = it.values + (id to applied)) }
            } catch (e: Exception) {
                eventChannel.send(e.message)
            } finally {
                _uiState.update { it.copy(updating = it.updating - id) }
            }
        }
    }

    companion object {
        val NETWORK_TOGGLES = listOf(
            PerfToggle("brotli", R.string.perf_brotli),
            PerfToggle("http2", R.string.perf_http2),
            PerfToggle("http3", R.string.perf_http3),
            PerfToggle("0rtt", R.string.perf_0rtt),
            PerfToggle("early_hints", R.string.perf_early_hints),
            PerfToggle("websockets", R.string.perf_websockets),
            PerfToggle("ipv6", R.string.perf_ipv6),
        )
    }
}
